src/cr/cube/dimension.py

```diff
@@ -6,10 +6,244 @@
 
 import numpy as np
 
-from cr.cube import ITEM_DIMENSION_TYPES
+from cr.cube.enum import DIMENSION_TYPE as DT
 from cr.cube.util import lazyproperty, memoize
 
 
+class _BaseDimensions(Sequence):
+    """Base class for dimension collections."""
+
+    def __getitem__(self, idx_or_slice):
+        """Implements indexed access."""
+        return self._dimensions[idx_or_slice]
+
+    def __iter__(self):
+        """Implements (efficient) iterability."""
+        return iter(self._dimensions)
+
+    def __len__(self):
+        """Implements len(elements)."""
+        return len(self._dimensions)
+
+    @lazyproperty
+    def _dimensions(self):
+        """tuple of dimension objects in this collection.
+
+        This composed tuple is the source for the dimension objects in this
+        collection.
+        """
+        raise NotImplementedError('must be implemented by each sublass')
+
+
+class AllDimensions(_BaseDimensions):
+    """Collection containing every dimension defined in cube response."""
+
+    def __init__(self, dimension_dicts):
+        self._dimension_dicts = dimension_dicts
+
+    @lazyproperty
+    def apparent_dimensions(self):
+        """_ApparentDimensions collection of the "visible" dimensions.
+
+        The two dimensions for a multiple-response (MR) variable are
+        conflated into a single dimensions in this collection.
+        """
+        return _ApparentDimensions(all_dimensions=self._dimensions)
+
+    @lazyproperty
+    def _dimensions(self):
+        """tuple of dimension objects in this collection.
+
+        This composed tuple is the internal source for the dimension objects
+        in this collection.
+        """
+        return tuple(
+            _DimensionFactory.iter_dimensions(self._dimension_dicts)
+        )
+
+
+class _ApparentDimensions(_BaseDimensions):
+    """Collection containing only "user" dimensions of a cube."""
+
+    def __init__(self, all_dimensions):
+        self._all_dimensions = all_dimensions
+
+    @lazyproperty
+    def _dimensions(self):
+        """tuple of dimension objects in this collection.
+
+        This composed tuple is the source for the dimension objects in this
+        collection.
+        """
+        return tuple(
+            d for d in self._all_dimensions
```

Comment:
`d` seems a bit too short for me personally. I know that linter used to complain, but I didn't check it now. It's super minor, you don't have to fix it, but just wanted to point out that it _might_ need to be fixed if we firm our position on code formatting.

Reply:
Well, my rationale here is that the scope is so very, very small (one line) and the contents of `._all_dimensions` is clear.

We could go with `dimension for dimension in self._all_dimensions`, but for whatever reason I find this expression clearer, like it would be in math "Let d represent a dimension". There's some tendency in functional programs to use very short names like these in a short scope (like 3 or 4 lines, no more), and I've been experimenting a little to see when I find they read better. I understand of course that readability is somewhat subjective at this level of detail.

I would suggest we let this one stand and see how we like it when returning back to it after a few weeks or whatever. It's difficult to distinguish 'different' from 'worse' at first, so if we age it a while and see what we think then we can make a change if we want.